sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,634 @@
+          replica.getCoreUrl(), replica.getStr(ZkStateReader.BASE_URL_PROP));
+    }
+  }
+
+  /** A {@link BoundedSource} reading from Solr. */
+  @VisibleForTesting
+  static class BoundedSolrSource extends BoundedSource<SolrDocument> {
+
+    private final SolrIO.Read spec;
+    // replica is the info of the shard where the source will read the documents
+    @Nullable private final ReplicaInfo replica;
+
+    BoundedSolrSource(Read spec, @Nullable Replica replica) {
+      this.spec = spec;
+      this.replica = replica == null ? null : ReplicaInfo.create(replica);
+    }
+
+    @Override
+    public List<? extends BoundedSource<SolrDocument>> split(long desiredBundleSizeBytes,
+        PipelineOptions options) throws Exception {
+      List<BoundedSolrSource> sources = new ArrayList<>();
+      int numShard;
+      try (AuthorizedCloudSolrClient client = spec.getConnectionConfiguration().createClient()) {
+        String collection = spec.getConnectionConfiguration().getCollection();
+        DocCollection docCollection = client.getDocCollection(collection);
+        numShard = docCollection.getSlices().size();
+        for (Slice slice : docCollection.getSlices()) {
+          sources.add(new BoundedSolrSource(spec, slice.getLeader()));
+        }
+      }
+      checkArgument(sources.size() == numShard, "Not enough leaders were found");
+      return sources;
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws IOException {
+      if (replica != null) {
+        return getEstimatedSizeOfShard(replica);
+      } else {
+        return getEstimatedSizeOfCollection();
+      }
+    }
+
+    private long getEstimatedSizeOfShard(@Nonnull ReplicaInfo replica) throws IOException {
+      try (AuthorizedSolrClient solrClient = spec.getConnectionConfiguration()
+          .createClient(replica.baseUrl())) {
+        CoreAdminRequest req = new CoreAdminRequest();
+        req.setAction(CoreAdminParams.CoreAdminAction.STATUS);
+        req.setIndexInfoNeeded(true);
+        CoreAdminResponse response;
+        try {
+          response = (CoreAdminResponse) solrClient.process(req);
+        } catch (SolrServerException e) {
+          throw new IOException("Can not get core status from " + replica, e);
+        }
+        NamedList<Object> coreStatus = response.getCoreStatus(replica.coreName());
+        NamedList<Object> indexStats = (NamedList<Object>) coreStatus.get("index");
+        return (long) indexStats.get("sizeInBytes");
+      }
+    }
+
+    private long getEstimatedSizeOfCollection() throws IOException {
+      long sizeInBytes = 0;
+      ConnectionConfiguration config = spec.getConnectionConfiguration();
+      try (AuthorizedCloudSolrClient solrClient = config.createClient()) {
+        DocCollection docCollection = solrClient.getDocCollection(config.getCollection());
+        for (Slice slice : docCollection.getSlices()) {
```

Comment:
How many slices can there be in practice? E.g. if there can be hundreds of thousands, then it might not be worth it to do hundreds of thousands of RPCs in getEstimatedSize(), which is called at pipeline submission time and could delay the pipeline submission considerably. If it can conceivably take more than a minute or so, it's most likely not worth it.

It's fine to leave as a TODO, but please at least add a comment saying whether this is a fast operation or it can take a long time (and then add a TODO to address it, e.g. only look at some slices rather than all of them, and extrapolate)